### PR TITLE
ARROW-3538: [Python] ability to override the automated assignment of uuid for filenames when writing datasets

### PR DESCRIPTION
**Pull request by Tomme**
Implementation for: [ARROW-3538](https://issues.apache.org/jira/browse/ARROW-3538).
Allowing one to overwrite datasets rather than always creating a new copy or appending.

### CODE REVIEW

**Comment by wesm** [comment]
Seems like the argument should be a callback function of some kind.

Note that this code is on death row pending a C++ rewrite

**Comment by wesm** [comment]
Are you open to making the argument a function? The function can be passed the partition keys or something similar

**Comment by Tomme** [comment]
> Are you open to making the argument a function? The function can be passed the partition keys or something similar

Sounds like a good idea to me, and happy to do. Feel free to close the PR and I can open a new one (or keep this open, I am not fussed).

**Comment by jorisvandenbossche** [comment]
> Feel free to close the PR and I can open a new one (or keep this open, I am not fussed).

I think you can update this PR, no need to open a new one.

**Comment by emkornfield** [comment]
@Tomme thank you for the work so far.  Do you think you will be able to update this soon?

**Comment by Tomme** [comment]
@emkornfield or anyone else, PTAL.

**Comment by emkornfield** [comment]
@tomme could you provide a unit test to verify functionality, otherwise seems reasonable to me (but I'm not an expert in this area of the code).

File: python/pyarrow/tests/test_parquet.py [hunk 1]
```diff
@@ -2301,6 +2301,64 @@ def test_write_to_dataset_no_partitions(tempdir):
...
+
+    # Partitioned columns become 'categorical' dtypes
+    input_df = input_df[cols]
+    for col in partition_by:
+        output_df[col] = output_df[col].astype('category')
+    assert output_df.equals(input_df)
+
+    # ARROW-3538: Ensure partition filenames match the given pattern
+    # defined in the local function partition_filename_callback
+    df_partition_keys = set(tuple(key)
```
Comment (emkornfield) [inline python/pyarrow/tests/test_parquet.py]: small nit:  I think it would be better to explicitly give the values you are expecting as strings instead of having logic here.  and call this expected_df_file_base_names and then don't do the split below.

File: python/pyarrow/tests/test_parquet.py [hunk 2]
```diff
@@ -2301,6 +2301,64 @@ def test_write_to_dataset_no_partitions(tempdir):
...
+    pq.write_metadata(output_table.schema, metadata_path)
+
+    # ARROW-2891: Ensure the output_schema is preserved when writing a
+    # partitioned dataset
+    dataset = pq.ParquetDataset(path,
+                                validate_schema=True)
+
+    # ARROW-2209: Ensure the dataset schema also includes the partition columns
+    dataset_cols = set(dataset.schema.to_arrow_schema().names)
+    assert dataset_cols == set(output_table.schema.names)
```
Comment (jorisvandenbossche) [inline python/pyarrow/tests/test_parquet.py]: General comment: I don't think it is needed to repeat the (for this case unrelated) asserts that are already in the other tests. I would also remove the references to the other ARROW-xxxx issues that are not related to ARROW-3538. 
I would just focus on testing the additional aspect of the file names.

**Review by jorisvandenbossche — APPROVED** [review]
Thanks for the update!

File: python/pyarrow/tests/test_parquet.py (hunk 1 above)
Comment (Tomme) [reply python/pyarrow/tests/test_parquet.py]: Good spot, I don't know why I made it more complicated than it needed to be.

File: python/pyarrow/tests/test_parquet.py (hunk 2 above)
Comment (Tomme) [reply python/pyarrow/tests/test_parquet.py]: Makes complete sense - was simply unsure if custom file names might cause any knock-on effects.

**Comment by emkornfield** [comment]
+1, thank you.

**Comment by emkornfield** [comment]
@Tomme can you provide your JIRA username or create a new account on https://issues.apache.org/jira/browse/ARROW-3538 so this can be assigned to you?

**Comment by Tomme** [comment]
> @Tomme can you provide your JIRA username or create a new account on https://issues.apache.org/jira/browse/ARROW-3538 so this can be assigned to you?

Sure thing, my username is `Tomme` on JIRA, I also added a comment if that helps.
